### PR TITLE
Fix computation of LL(k) FIRST, FOLLOW, LOOKAHEAD

### PR DESCRIPTION
**Pull request by apalala**
Computation of FIRST, FOLLLOW, and LOOKAHEAD sets was originally implemented to provide the parser and parser generators with information about expected tokens on parse failures.

The implementation was incomplete, and buggy. This set of changes fixes that.

Additionally, it adds `ref(rulename)` tokens to the LOOKAHEAD sets to allow the algorithm to work with left recursive grammars.

There's a chance that this implementation will lead to much simpler computations of `is_leftrec`, `is_nullable`, and `is_memoizable`.

### CODE REVIEW

**Comment by norswap** [comment]
Hey Juancarlo!

I don't have the time (nor really the familiarity) to review the code, but I can comment on what you wrote on the PEG mailing list with some thoughts:

> When looking at the resolutions of PEG/leftre with IS_NULLABLE, IS_LEFTREC, it occured to me that the same could be solved, perhaps in a more algebraic way, by using LL FIRST/FOLLOW analysis with the modification of allowing RHS symbols in the sets.
> 
> IS_NULLABLE(rule) == e in pFIRST(rule)
> IS_LEFTREC(rule). == rule inf pFIRST(rule)

That first equation looks very shaky to me. It's not enough to have the empty string (epsilon) as a potential first expression. The issue is sequences. For instance: `X ::= a? b` has epsilon in FIRST, but is not nullable, since it has to at least match `b`.

Regarding left-recursion, that does indeed work nicely.

The way I formalize left-recursion in my thesis is that I say there exists a relationship `FIRST(A, B)` between the parsing expression `A` and `B` if `B` can be **directly** invoked by `A` at the same input position it was itself invoked at. This relationship gives you a directed graph and left-recursion is a loop in the graph.

What you're doing here is defining your `FIRST(A)` set as the transitive closure of `FIRST(A, B)` relationship with a fixed `A`. Or from the graph perspective, you're computing the set of reachable expressions from A. If A is reachable from itself, you have a loop, hence left-recursion.

---

Regarding your sets, I understand `FIRST` and `FOLLOW`, but what is the semantics of `LOOKAHEAD`? Is it like `FIRST`, but within a lookahead (positive only, or negative as well)?

To avoid left-recursion, you have to check all cases (actually invoked at the same position without lookahead, with positive lookahead, with negative lookahead).

**Comment by apalala** [comment]
@gvanrossum I didn't call your attention to actually review the code, but just so you know there may be a much simpler way to implement PEG/lefrec by computing FIRST sets (something you're familiar with) augmented by inclusion of no-terminals (rule names). The work to know if performance of PEG with a C implementation is good enough for Python remains pending.

**Comment by apalala** [comment]
(Copied from peg@mit.)

The first rule should have been:

IS_NULLABLE(exp) == e in pFIRST(exp) 

When applied to rules, exp should be the right hand side, which may be a sequence.

In your example, `X ::= a? b`, `?a` is nullable, but `a? b` shuld not be as it can never generate the empty sequence. The FIRST set should be {a, b}. I'll add this example as a unit test.
Ah! (Re: your Thesis work) I knew this must had been thought about before. In my proposal, pFIRST(exp) may include non-terminals (which don't interfere with terminals, and can safely be discarded if working towards LL(1) )

LOKAHEAD is the k-concatenation of FIRST and FOLLOW in LL(k), the predictive k in LL(k). It is needed precisely to deal with rules that are nullable. I can't remember an example for the need off the top of my head. 

My interest in this approach is three-fold:
* The theoretical basis is much clearer because it's algebraic/graph-based (as you explain above)
* Computation of FIRST/FOLLOW is well understood and there are simple algorithms for it (like the one in TatSu)
* The above makes it easier to implement PEG/leftrec in lower level languages (like in C, which must be the target in my work to help with next-generation Python compilation).

**Comment by norswap** [comment]
I don't think the confusion is with regards to rules vs expressions (I don't really understand how it changes anything actually, but this is probably just an issue of definition).

The issue is that I understood `FIRST(exp)` to be a set of expressions (where `epsilon` is the parsing expression matching nothing). You seem to understand it as a set of terminals + `epsilon`.

At that point the question becomes "what does it mean to have `epsilon` in `FIRST(A)` ?", and I really don't see what your answer to that question is.

Because my current understanding is: we add `epsilon` to the set if the expression is nullable. Which is cyclical and makes the `IS_NULLABLE(exp) == e in pFIRST(exp)` definition useless.

And this ties in to your comment on `LOOKAHEAD`. If this set is needed for the computation of nullability, it **should** appear in the definition!

And indeed, it's exactly the problem I point out: for a sequence to be nullable, every single item in the sequence must itself be nullable.

---

Micro rant:

I think people are over obsessed with theory, when they should be obsessed with simplicity. If it's simple, you can make a simple theory of it.

And sometimes, a slightly more complex, yet more understandable theory, is better.

---

Regarding nullability, the way I compute it in Autumn is basically as a set of recursive equations, one for each type of parsing expression.

So you have:
- `NULLABLE(seq)` <= every item in `seq` is nullable
- `NULLABLE(choice)` <= at least one item in `choice` is nullable
- `NULLABLE(optional)` <= true
- `NULLABLE(character)` <= false

etc. and it's just implemented as a set of recursive functions
https://github.com/norswap/autumn4/blob/master/src/norswap/autumn/visitors/_VisitorNullable.java

---

Just to clear up any confusion, the graph approach doesn't involve "non-terminals" at all. In fact non-terminals aren't even "reified" in Autumn, they are just labels on parsing expressions which form a de-facto graph (e.g. a sequence node has an edge going to all its children).

Similarly the graph induced by `FIRST(A, B)` is also de facto, I don't actually have a graph data structure.

In fact I do it just like you: I computer the `FIRST` set (although for me that is a set of parsing expressions, not a set of terminals):

https://github.com/norswap/autumn4/blob/master/src/norswap/autumn/visitors/_VisitorFirstParsers.java

Then it is used to find left-recursion:

https://github.com/norswap/autumn4/blob/master/src/norswap/autumn/visitors/WellFormednessChecker.java

**Comment by gvanrossum** [comment]
@norswap: I have a similar way to compute NULLABLE using a set of recursive functions, but there's still the issue of what to do with left-recursive rules. If we have
```
expr: expr '+' term
    | term
```
then asking NULLABLE(expr) is recursive. I detect the recursion, but then what should I return? (I don't want to use left-recursion in the definition of NULLABLE, since NULLABLE is needed for the definition of left-recursiveness.) In this, and most realistic examples, returning False (at least for the left-recursive alternative) seems the right answer, but is it always?

@apalala: While I haven't got indirect left-recursion working, I would like my next focus to be generating C code that can be slotted into CPython to replace the pgen-based parser and ast.c. I may be able to make some progress with that at the next core-dev sprint (September), or over the summer.

**Comment by norswap** [comment]
@gvanrossum I actually keep a stack of traversed parser to detect/cutoff recursion.

(Everything goes through this function here: https://github.com/norswap/autumn4/blob/master/src/norswap/autumn/visitors/_VisitorNullable.java#L59-L74)

So it's not actually as simple as recursive functions, but it looks as though it is.

Recursive invocations (left, right, no matter) are treated as non-nullable — ultimately this means that the recursive rule itself will have the same nullability as its non-recursive base cases, which is what we want.

**Comment by apalala** [comment]
I agree with your comment about "Occam's Razor", @nicholasbishop. In fact, I think that the common case is that the simpler solution proceeds the simpler theory.

If the empty sequence belongs to an LL(k) language, then `ε` must be present in the LOOKAHEAD set. That is usually resolved  by including in compiler books by including `$` as a terminal symbol in new start rule `S1 -> S $`.

My argument is that your definitions for `NULLABLE` and `IS_LEFTREC` are equivalent to mine under the semantics of `k-concatenation` (`•k`) of LL(k). Given two sets of sequences/strings of symbols (that may include `ε`, the empty sequence), `•k` will produce sequences cut to at most `k` symbols of:

* the other set if one of the input sets is empty
* the concatenation of the sequences in the cartesian product of the input sets

Since `•1` is enough for our purposes, you have:
* `NULLABLE(seq)` if `ε in FIRST(seq)`, which is defined as `ε in (FIRST(s0) • ... • FIRST(sn))` which will happen only if `ε` is in all `FIRST(si)`, which by definition will only happen if each set is `NULLABLE`
* NULLABLE(choice) is `ε in (FIRST(c0) | ...| FIRST(cn))` which will happen only if `ε` is in one of the sets, IOW, if one of the choices is `NULLABLE`
* `NULLABLE(optional)` is `ε in FIRST(e?) == ε in ({ε} | FIRST(e))`, which is always true 
* NULLABLE(character) is `ε in FIRST(character) == ε in {character}` which is always false.

The definition of `pFIRST(e)` is:
* `FIRST(e) | {e}` if `e` is a non-terminal
* `FIRST(e)` otherwise
  
There doesn't need to be a definition of `NULLABLE` to prove that:

* `IS_LEFTREC(rule) == rule in pFIRST(rule-right-hand-side)`

And that is enough even for indirect left-recursion. A difference is that "intermediate" rules currently marked as just "not memoizable" in the current implementation in TatSu will now be marked as recursive, which to me makes more since, since a parse started on any of the involved rules _will_ be recursive.

And that reminds me that @gvanrossum has a _very nice_ implementation of left-recursion parsing for PEG _within_ the memoizer in `pegen` (that I plan to ~steal~ imitate). 

P.S. I used `pLOOKAKEAD` in my previous examples trying to be complete, as concatenation with `pFOLLOW` is intended to take care of rules like `A-> ε` in LL(k) (there's a nice explanation [here](https://stackoverflow.com/a/20317336/545637), but I think `pFIRST` is enough (I'll eventually need to prove that `pLOOKAHEAD` may erase `NULLABLE` but not `IS_LEFTREC`, which is what we're ultimately after).

**Comment by apalala** [comment]
Yes, @gvanrossum , the C implementation for Python/PEG is the priority. I'm still taking time off _some_ weekends to work on the plan. To keep sharing what we've found is the fastest way one of us to get there.

**Comment by gvanrossum** [comment]
> And that reminds me that @gvanrossum has a _very nice_ implementation of left-recursion parsing for PEG _within_ the memoizer in `pegen` (that I plan to ~steal~ imitate).

I've made the pegen repo public, so here's a [pointer to the code](https://github.com/gvanrossum/pegen/blob/400a489404aa8297c0a9619b2ea7262461abfba5/pegen.py#L173-L209) you mentioned. Do note that I got the idea from [Philippe Sigaud's Pegged wiki](https://github.com/PhilippeSigaud/Pegged/wiki/Left-Recursion).

Note that I haven't figured out how to do this for indirect left-recursion yet.

**Comment by apalala** [comment]
This fixes #138

**Comment by apalala** [comment]
> Note that I haven't figured out how to do this for indirect left-recursion yet.

I read the comments in your implementation, and that's why I brought you here, @gvanrossum .

The algorithm discussed in this PR detects indirect left-recursion in a simple way (see [here](https://github.com/neogeny/TatSu/blob/fix_first_follow/tatsu/grammars.py#L41) for the key). I think that all you need to do in your implementation is to give the same treatment to all rules involved in a leftrec cycle.

Yet, regarding Python, a proof of concept in C (not leftrec) remains the greatest priority.

**Comment by norswap** [comment]
@gvanrossum I can suggest my own approach for left-recursion, which I think is a lot simpler than what is proposed elsewhere.

It's outlined in the paper here: https://norswap.com/pubs/sle2015.pdf but I'll give a terse explanation below.

(Fair warning, the part of the algorithm that deals with left-associativity is slightly broken - but if you don't implement explicit associativity selection, you don't need it.)

The approach has two parts.

First, you determine a set of parsing expressions to mark as left-recursive. This is done by walking the graph generated by the `FIRST(A, B)` relationship, where `A` and `B` are both parsing expressions and `A` can directly invoke `B` at the same position that `A` was itself invoked at. For this, you'll need a `NULLABLE(A)` predicate. Anyway, you find cycles in this graph and you select a node in the cycle to mark as left-recursive (typically the first you encounter, which is a parsing expression corresponding to a named grammar rule).

(I can supply more details if required.)

Once that is done, you replace all these left-recursive nodes by left-recursive expressions. which wrap the original node.

Regarding the code of that expression, here is a very quick Python prototype:

```python
seeds = {}

class ParsingExpression (object):


    def parse (self, input: str, position: int) -> (bool, int):
        """Takes the input and the current position and return a pair
        whose first item indicates whether the parsing expression
        succeeded in matching at that position, and whose second item
        indicates the input position past the match in case of success.
        """
        raise NotImplementedError("please implement")

class LeftRecursive (ParsingExpression):
    
    def __init__ (self, child: ParsingExpression):
        self.child = child
        
    def parse (self, input: str, position: int) -> (bool, int):
        seed = seeds.get((position, self), None)
        if seed != None:
            return seed
        current = (False, position)
        seeds[(position, self)] = current
        while True:
            success, new_position = self.child.parse(position)
            if success and new_position > current.second:
                current = (success, new_position)
                seeds[(position, self)] = current
            else:
                del seeds[(position, self)]
                return current
```

The behaviour you get is that left-recursive rules simply work. The only contentious case is rules that are both left- and right-recursive (e.g. `A ::= A + A` — afaik this can always be avoided). In this case, it still works but you'll always get a right-associative parse. You need further changes to the algorithm to be able to specify a left-associative parse.

I can supply more clarifications and information if needed!

**Comment by gvanrossum** [comment]
Yup, that is almost literally what I had. The one tip I get from your explanation is that I only need to mark one rule in a cycle as left-recursive.

**Comment by norswap** [comment]
@gvanrossum Good to hear! Managed to fix the issue with indirects then?

A point of precision: the algorithm does work when multiple nodes in a cycle are wrapped in a left-recursive node (in fact this is sometimes necessary when you have nested cycles). However, this introduces undesirable overhead. Assuming every `n` nodes of a cycle are made left-recursive, the overhead is of the order of `2^n` operations.

**Comment by gvanrossum** [comment]
I have it working for some very simple indirects, which is all I need for now. I only wrap the first node, and the others are not memoized (that's my new trick).

**Comment by norswap** [comment]
@gvanrossum Ah! Yes, my system assumes no memoization, and in fact you can't memoize nodes in the cycles or you won't get correct results.

However, if no nodes in the cycle are memoized, then indirects shouldn't be an issue. Since the mechanism at play is just repeating the parse, the algorithm doesn't even know whether recursion is direct or indirect.

Might be an issue in how the parsing expression graph is setup? The last expression in the cycle should refer back to the left-recursive expression (not to the original one that is wrapped inside)? I'd need to look at the code to know more, but I'm wrapping up my thesis right now so I don't really have the time ^^

On the topic, of memoization, it's not clear really speeds up parses when you factor out problematic ways to parse infix expressions (unfortunately a thorny issue, I have a lot to say on the topic). It might depend on the underlying language implementation too (the only way to know is to measure).

**Comment by gvanrossum** [comment]
- My solution combines the memoization logic with the (your?) algorithm for handling left-recursion -- your `seed` dict is my memo dict, essentially. Making other rules in the cycle not memoized solves the issue, as long as the cycle always goes back to the one rule that I call the "leader" (which has the left-recursive memoization decorator (`@memoize_left_rec`).
- I believe that because of this, having multiple rules in a cycle use `@memoize_left_rec` (i.e. multiple "leaders") doesn't work -- experimentally it doesn't, but I haven't debugged why.
- I think I can handle rules whose `FIRST(A, B)` subgraph looks like `A -> B -> C -> A`, but I think it will fail if there's also an arc `C -> B`, assuming `A`is the leader.
- But I don't see a need for such anomalies in a grammar for Python.
- I recall doing something similar to what I read in your thesis with infix operators in the early '80s, for ABC (Python's predecessor, designed by Lambert Meertens). Something like `A + B * C - D` would initially be parsed into an array containing `A <op> B <op> C <op> D` and then a separate algorithm would do some kind of shift-reduce thing on that taking the operator priorities into account, e.g. if its stack top was `A + B` and the next operator was `*` then it would do a shift (push `*` on the stack, and then `C` after that), but if the next operator was `+` it would do a reduce (combine `A + B` into a single `Add(A, B)` node). The extra-interesting thing was that ABC had "fuzzy" priorities so some operators would error out when adjacent (forcing the user to parenthesize certain sub-expressions where otherwise ambiguity for humans would arise). It was all done in an ad-hoc recursive-descent parser without backtracking, IIRC.

**Comment by norswap** [comment]
So I'm guessing reaching the leader temporarily disables memoization until the call stack unwinds at the leader?

Regarding left-rec + memo, I know that the Ohm parser does it (https://github.com/harc/ohm). It evolved from OMeta who had the first left-recursive algorithm — it relied on the memo store. I've seen aspersions cast on the published versions multiple times (haven't investigated), but that's doubtlessly fixed in Ohm.

Indeed, nested cycles are very nasty cases that are rarely useful. In fact, my own conclusion (which apparently you might have seen already :)) is that left-recursion is best avoided totally. You only really need them for infix expressions, and can design custom combinators for that.

The thesis currently misses the next step of that whole reasoning though. Which is that even with the folding combinator, you still run into performance issues with operators at the same precedence level (at least without memoization), so I have a combinator that lets you define a precedence level at once. Seems to be the more UX-friendly option so far!

ABC: Nice! Is there some kind of document / artifact on that somewhere? I wouldn't be surprised that this has been reinvented time and time again. I know at least some cases where AST rewriting is being done to the same effect.

**Comment by gvanrossum** [comment]
ABC: https://homepages.cwi.nl/~steven/abc/ -- there's a link to
implementations there that has a zip of the original source code (C). I
thought there was a GitHub mirror but I can't find it now. There's also
https://github.com/ramalho/abc which seems to be an archive of original ABC
papers etc.

I don't know what the performance of my approach will be yet -- I'd like to
change the generator to produce C code and compare that to the performance
of the current CPython parser. But it'll take all summer at my current pace.

I'm not quite as down on left-recursion as you are -- I still like that it
maps the grammar directly to the ABC that Python uses as input to its
compilation phase. But I'll keep an eye out for alternatives.

**Comment by ltratt** [comment]
> Regarding left-rec + memo, I know that the Ohm parser does it (https://github.com/harc/ohm). It evolved from OMeta who had the first left-recursive algorithm — it relied on the memo store.

My memory on this is a bit out of date, but you don't need memoisation to implement the left-recursive algorithm (see Sec 4.2 of https://tratt.net/laurie/research/pubs/html/tratt__direct_left_recursive_parsing_expression_grammars/). That said, IMHO, the approach (whether memoised or not) leads to surprising parses (see Sec 5 of that paper). The Lua folk have an approach which they claim can solve this problem (Sec 4 of https://www.sciencedirect.com/science/article/pii/S0167642314000288), but I must admit that I'd moved on from this stuff at that point, so I don't think I ever acquired a good understanding of how it might work.

**Comment by norswap** [comment]
@ltratt We know, the algorithm I posted above works for non-memoizing parsers :) It also work for indirect and hidden left-recursion. Really the only restriction is that there *shouldn't* be memoization used there.

I looked into Medeiros et al. for my thesis. The thing about their approach is that it relies on labeling non-terminals with priorities and they dismiss that work as "something a tool can do". IIRC this can work if you want the default right-associative interpretation. If you want to enable left-associativity though, it's actually impossible in general because of hidden left/right-recursion: some rules might left-recurse or right-recurse or not depending on the input. If you constraint yourself to a well-defined set of patterns, it can work. I argue against that in the thesis, though in favor of a simpler solution (but still left-recursive with associativity selection).

Basically, as soon as you start putting associativity into left-recursion, the semantics become very muddy.
